### PR TITLE
fixing namespaces,

### PR DESCRIPTION
**Pull request by mohsenShakiba**
added c# 9.0 language for TrottlR,
better naming for variables,
added binary serializer for counter,
general improvements

### CODE REVIEW

File: sample/MVC/Startup.cs [hunk 1]
```diff
@@ -2,7 +2,10 @@
 using Microsoft.AspNetCore.Builder;
 using Microsoft.AspNetCore.Http;
 using Microsoft.Extensions.DependencyInjection;
-using ThrottlR;
+using ThrottlR.DependencyInjection;
+using ThrottlR.Endpoints;
+using ThrottlR.Middleware;
+using ThrottlR.Policy;
```
Comment (Kahbazi) [inline sample/MVC/Startup.cs:8]: Having only `ThrottlR` for namespace in intentional, so the client code doesn't need to have multiple using statements.

File: src/ThrottlR/DependencyInjection/ThrottlerServiceCollectionExtensions.cs [hunk 2]
```diff
@@ -58,9 +61,19 @@ public static IThrottlerBuilder AddInMemoryCounterStore(this IThrottlerBuilder b
             return builder;
         }
 
-        public static IThrottlerBuilder AddDistributedCounterStore(this IThrottlerBuilder builder)
+        public static IThrottlerBuilder AddDistributedCounterStore(this IThrottlerBuilder builder,
+            bool useStringSerializer = false)
         {
-            builder.Services.TryAddSingleton<ICounterStore, DistributedCacheCounterStore>();
+            builder.Services.AddDistributedMemoryCache();
```
Comment (Kahbazi) [inline src/ThrottlR/DependencyInjection/ThrottlerServiceCollectionExtensions.cs:67]:
```suggestion
```
`AddDistributedCounterStore` works on top of `IDistributedCache`, and the client must determine what type of distributed cache it needs.

File: src/ThrottlR/ThrottlR.csproj [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
   <PropertyGroup>
     <TargetFramework>netcoreapp3.1</TargetFramework>
+    <LangVersion>9.0</LangVersion>
```
Comment (Kahbazi) [inline src/ThrottlR/ThrottlR.csproj:5]:
```suggestion
```
I don't see any critical reason to upgrade the language version to 9.0 right now, also I haven't update the build pipeline yet. This is why the build fails. 
https://github.com/Kahbazi/ThrottlR/pull/1/checks?check_run_id=1463371243#step:5:9

**Review by Kahbazi — CHANGES_REQUESTED** [review]
Thank you for taking the time to send the very first PR on this project.

I reviewed the code. I have a different opinion on the namespace for the classes and folder structures. I like them the way they are now. I see no reason to change them now.

Also In order to better manage the PRs, I recommend only focus on one issue per PR.

I like the idea to use binary serializer in `DistributedCacheCounterStore`. Could you please only keep the changes in regards to binary serializer? Also there's no need to introduce new type for `DistributedCacheCounterStore`. Changing its implementation is good enough.

File: src/ThrottlR/DependencyInjection/ThrottlerServiceCollectionExtensions.cs (hunk 2 above)
Comment (mohsenShakiba) [reply src/ThrottlR/DependencyInjection/ThrottlerServiceCollectionExtensions.cs:67]: sorry, it slipped through

**Comment by mohsenShakiba** [comment]
I see, could you point me to the any resource about how you prefer the structure of .net project, because the default convention of namespaces is quite different compared to what I saw. I don't want force you to change your structure but instead find out about your reasoning, if it's just something that you prefer without any backing sources then it's totally cool.

a couple of questions though:
- why using ` await Task.Delay(2000);` it makes the benchmark test slow?
- should i keep the `ThrottleRule` as struct?
- the `ProcessRequestAsync` method on `ThrottlerService` has an unnecessary branch, should I keep my changes?
- I changed the way that `GenerateCounterKey` works on `ThrottlerItem` it had an unnecessary branch, should I keep my changes?

last but not least I spent a good half and hour tinkering with `ThrottlerService->GetRules` and I still have no clue on how it works, do you think that you can add some comments when you get to it, it's a bit confusing.

but any way I'll do as you said and I've already learned a ton from your code.

**Comment by Kahbazi** [comment]
>I see, could you point me to the any resource about how you prefer the structure of .net project, because the default convention of namespaces is quite different compared to what I saw. I don't want force you to change your structure but instead find out about your reasoning, if it's just something that you prefer without any backing sources then it's totally cool.

For extensions methods I like to have the same namespace as the type I'm writing the extension on, so it can be discoverd easier.
For the other parts it's more personal preference at this point, but it might change in future.

> why using await Task.Delay(2000); it makes the benchmark test slow?

I was trying different delays to see the results, this could change to read from a property with `Params` attribute.

> should i keep the ThrottleRule as struct?

I don't see any reason not to. Create a PR for it. I will check the potential side effects.

> the ProcessRequestAsync method on ThrottlerService has an unnecessary branch, should I keep my changes?
> I changed the way that GenerateCounterKey works on ThrottlerItem it had an unnecessary branch, should I keep my changes?

Can you create separate issues for these and explain the unnecessary branch. 

>last but not least I spent a good half and hour tinkering with ThrottlerService->GetRules and I still have no clue on how it works, do you think that you can add some comments when you get to it, it's a bit confusing.

I agree. The project needs documents. There are two rules, general and specific. General rules apply to all and specific apply based on scope (ip, username, ...). A rule has a time window and a quota. Specific rules would override a general rule with the same time window. Perhaps checking the tests makes it more clear. https://github.com/Kahbazi/ThrottlR/blob/2b0e16f6d7b95a4a64c12483db34a5f98ef7b680/test/ThrottlR.Tests/ThrottlerServiceTests.cs

**Pull request closed without merging**